rfc/rfc-83/rfc-83.md

```diff
@@ -0,0 +1,173 @@
+
+This RFC will draw on the design of Incremental Clean to generalize the capability of processing incremental partitions to all table services, such as Clustering and Compaction.
+
+## Background
+
+`earliestInstantToRetain` in clean plan meta
+
+HoodieCleanerPlan.avsc
+
+```text
+{
+  "namespace": "org.apache.hudi.avro.model",
+  "type": "record",
+  "name": "HoodieCleanerPlan",
+  "fields": [
+    {
+      "name": "earliestInstantToRetain",
+      "type":["null", {
+        "type": "record",
+        "name": "HoodieActionInstant",
+        "fields": [
+          {
+            "name": "timestamp",
+            "type": "string"
+          },
+          {
+            "name": "action",
+            "type": "string"
+          },
+          {
+            "name": "state",
+            "type": "string"
+          }
+        ]
+      }],
+      "default" : null
+    },
+    xxxx
+  ]
+}
+```
+
+`EarliestCommitToRetan` in clean commit meta
+
+HoodieCleanMetadata.avsc
+
+```text
+{"namespace": "org.apache.hudi.avro.model",
+ "type": "record",
+ "name": "HoodieCleanMetadata",
+ "fields": [
+     xxxx,
+     {"name": "earliestCommitToRetain", "type": "string"},
+     xxxx
+ ]
+}
+```
+How to get incremental partitions during cleaning
+
+![cleanIncrementalpartitions.png](cleanIncrementalpartitions.png)
+
+**Note**
+`EarliestCommitToRetain` is recorded in `HoodieCleanMetadata`
+newInstantToRetain is computed based on Clean configs such as `hoodie.clean.commits.retained` and will be record in clean meta as new EarliestCommitToRetain
+
+## Design And Implementation
+### Abstraction
+Use `IncrementalPartitionAwareStrategy` to control the behavior of getting partitions, filter partitions, also incremental getting partitions if necessary etc.
+
+```java
+package org.apache.hudi.table;
+
+import org.apache.hudi.common.engine.HoodieEngineContext;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.config.HoodieWriteConfig;
+
+import java.util.List;
+
+public interface IncrementalPartitionAwareStrategy {
+
+  /**
+   * Get partition paths to be performed for current table service.
+   * @param metaClient
+   * @return
+   */
+  List<String> getPartitionPaths(HoodieWriteConfig writeConfig, HoodieTableMetaClient metaClient, HoodieEngineContext engineContext);
+
+  /**
+   * Filter partition path for given fully paths.
+   * @param metaClient
+   * @return
+   */
+  List<String> filterPartitionPaths(HoodieWriteConfig writeConfig, List<String> partitionPaths);
+
+  /**
+   * Get incremental partitions from EarliestCommitToRetain to instantToRetain
+   * @param instantToRetain
+   * @param type
+   * @param deleteEmptyCommit
+   * @return
+   * @throws IOException
+   */
+  List<String> getIncrementalPartitionPaths(HoodieWriteConfig writeConfig, HoodieTableMetaClient metaClient);
```

Comment:
Should we just add one interface `List<String> filterPartitionPaths(HoodieWriteConfig writeConfig, List<String> allPartitionPaths, List<String> incrementalPartitionPaths);` to that the strategy can decide which partition are choosed.

The `getXXXPartitionPaths` should belong to the scope of the executor/planner, let's move them out.